script/tool/lib/src/update_excerpts_command.dart

```diff
@@ -23,232 +26,302 @@ class UpdateExcerptsCommand extends PackageLoopingCommand {
-        .toList();
-    if (relativeMdPaths.isEmpty) {
-      return true;
+  _UpdateResult _updateExcerptsIn(File file) {
+    bool detectedChange = false;
+    List<String>? errors;
+    void reportError(String message) {
+      errors ??= <String>[];
+      errors!.add(message);
     }
-    final int exitCode = await processRunner.runAndStream(
-        'dart',
-        <String>[
-          'run',
-          'code_excerpt_updater',
-          '--write-in-place',
-          '--yaml',
-          '--no-escape-ng-interpolation',
-          ...relativeMdPaths,
-        ],
-        workingDir: example.directory);
-    return exitCode == 0;
-  }
 
-  /// Adds `code_excerpter` and `code_excerpt_updater` to [package]'s
-  /// `dev_dependencies` using path-based references to the submodule copies.
-  ///
-  /// This is done on the fly rather than being checked in so that:
-  /// - Just building examples don't require everyone to check out submodules.
-  /// - Examples can be analyzed/built even on versions of Flutter that these
-  ///   submodules do not support.
-  void _addSubmoduleDependencies(RepositoryPackage package,
-      {required Directory repoRoot}) {
-    final String pubspecContents = package.pubspecFile.readAsStringSync();
-    // Save aside a copy of the current pubspec state. This allows restoration
-    // to the previous state regardless of its git status at the time the script
-    // ran.
-    package.directory
-        .childFile(_originalPubspecFilename)
-        .writeAsStringSync(pubspecContents);
-
-    // Update the actual pubspec.
-    final YamlEditor editablePubspec = YamlEditor(pubspecContents);
-    const String devDependenciesKey = 'dev_dependencies';
-    final YamlNode root = editablePubspec.parseAt(<String>[]);
-    // Ensure that there's a `dev_dependencies` entry to update.
-    if ((root as YamlMap)[devDependenciesKey] == null) {
-      editablePubspec.update(<String>['dev_dependencies'], YamlMap());
+    String pathBase = path.dirname(file.path);
+    final StringBuffer output = StringBuffer();
+    final StringBuffer existingBlock = StringBuffer();
+    String? language;
+    String? excerpt;
+    _ExcerptParseMode mode = _ExcerptParseMode.normal;
+    int lineNumber = 0;
+    for (final String line in file.readAsLinesSync()) {
+      lineNumber += 1;
+      switch (mode) {
+        case _ExcerptParseMode.normal:
+          if (line.contains(_pragma)) {
+            RegExpMatch? match = _basePattern.firstMatch(line);
+            if (match != null) {
+              pathBase = match.group(1)!;
+            } else {
+              match = _injectPattern.firstMatch(line);
+              if (match != null) {
+                final String excerptPath = match.namedGroup('path')!;
+                final File excerptSourceFile = file.fileSystem.file(
+                    path.normalize(path.join(
+                        path.dirname(file.path), pathBase, excerptPath)));
```

Comment:
I'm confused how this works; aren't the first two things here *both* `pathBase`? (But this would become `Directory` operations, per the above.)

Reply:
the pathBase can be relative (in fact come to think of it it should originally be the empty string, rather than `path.dirname(file.path)`).

I'm happy to convert this to using Directories instead of paths but I'm not quite sure how I would handle relative path bases in that world. Can you elaborate?

Reply:
It's all relative to the `file` parameter, right? So you already have a `File`, and can do everything relative to `file.parent`, unless I'm missing something.

Reply:
Oh, I see. It wasn't intuitive to me that `directory.childFile` would take a relative path. As in, `directory.childFile('../../foo/bar.test')`. Seems like `childFile` is a weird name for that. :-)

Reply:
updated code as suggested

Reply:
only remaining use of `path` is for `path.extension`.